core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -196,11 +197,12 @@ private void rebuildIndex() throws Exception {
          int keyLen = buf.getInt();
          int dataLen = buf.getInt();
          int metadataLen = buf.getInt();
+         int internalMetadataLen = buf.getInt();
```

Comment:
I agree the MAGIC should be changed... but does it need to be backward compatible? 
isn't that the StoreMigrator's job?

Reply:
The `StoreMigrator` is a get out of jail, not quite for free, card. The long term goal is to get so that we can do rolling upgrades seamlessly in kubernetes, so we need no manual steps.

Reply:
oh... that was unexpected... 
SingleFileStore can be fixed but the remaining stores (RocksDB, JDBC, SIFS, JPA) are probably broken. there is no "magic" or "version" field there :/

Reply:
@pruivo we can relatively easy add a new column family for metadata in RocksDB, and a metadata table in JDBC/JPA. For SIFS we may need a breaking change.

Reply:
@danberindei not sure if it is going to work. JDBC and RockDB store the `MarshalledValue` as `byte[]`. I guess protostream will fail to unmarshall it because it expects to have another field... 

About JPA, my knowledge is -infinity. I just copy what others did
```
@Lob
@Column(length = 65535)
private byte[] internalMetadata;
```

Reply:
My bad, I was only thinking about the missing magic/version fields, so was using "metadata" as "information about the store itself", not as in "information about the entries".

I'm not very familiar with JPA either, but for JDBC/RocksDB it should be fairly easy to check when opening a store if the new table/column family exists or not.

However, this doesn't completely solve our problem, we also have to migrate the existing entries, or we wouldn't be able to change the version of the store, and we wouldn't be able to store the internal metadata.

Reply:
@pruivo @danberindei  For store implementations that use `MarshalledValueImpl` to store the value, i.e. Jdbc & Rocksdb, we should be ok because you have added the internal metadata there. The way Protobuf/ProtoStream works, if a field is missing from the old bytes, then it just isn't read and so in `MarshalledValueImpl` we can just set the internal metadata to null. So there shouldn't be any additional logic in the store itself AFAICT.

Reply:
> About JPA, my knowledge is -infinity. I just copy what others did

That's what we all do :laughing: